### ISSUE TITLE
Use ccache (optionally) in CMake builds

### ISSUE THREAD

**Issue by benjamin-bader** [Microsoft/ChakraCore#4775]
Hi folks,

Have you considered using ccache to speed up repeat builds on linux and mac machines?  It caches built object files, allowing them to be reused if the underlying source hasn't changed.  For us, this speeds our CI builds dramatically - we're building ChakraCore as a static library, and adopting ccache has taken our build from 30 minutes to one.

I'm proposing adding the following to the root CMakeLists.txt:

```cmake
if(USE_CCACHE) # set by build.sh
  find_program(CCACHE_PROGRAM ccache)
  if(CCACHE_PROGRAM)
    set(CMAKE_CXX_COMPILER_LAUNCHER ${CCACHE_PROGRAM})
    set(CMAKE_C_COMPILER_LAUNCHER   ${CCACHE_PROGRAM})
  endif()
endif()
```

Would you be amenable to such a change?

**Comment by jackhorton** [comment]
I don't see the issue with making a PR for this, but I also am a bit confused, since I usually don't have rebuild-the-world events in build.sh unless I change a core flag, like --no-icu to --system-icu (which changes some global defines). Could you provide a bit more info on how you are invoking build.sh each time?

**Comment by benjamin-bader** [comment]
Sure.  I'm incorporating ChakraCore as a CMake external project; a fresh build will clone the Chakra repo and build it on the dev's machine.  We do have build-the-world events occasionally, in the process of development.

It's more important for CI, where I definitely would prefer a clean build for each change.

**Comment by jackhorton** [comment]
Ah, that makes sense. I think this sounds pretty reasonable, though I don't think we would be able to take advantage of it ourselves since I think our CI runs don't share any state at all. @dilijev thoughts?

**Comment by dilijev** [comment]
You are right, no benefit to our CI, but embedders of ChakraCore may benefit from an improved experience / build times, so I don't see why not.

**Comment by obastemur** [comment]
IMHO we don't need `cache`. Current build script will rebuild only the files updated ++ you can set the output path as any location. So, re-clone wouldn't break the cache.

See `build.sh`
``
 --target-path[=S] Output path for compiled binaries. Default: out/
```

**Comment by benjamin-bader** [comment]
The benefit accrues primarily to CI scenarios, where we strongly prefer a clean build each time.  Having a compiler cache means that such changes add maybe 1-5 minutes to the build, rather than 35.

**Comment by jackhorton** [comment]
I am unfamiliar with CCache and a quick glance at the website doesn't tell me a whole lot, but what exactly does it do that --target-path doesn't do? SCCache's benefit is more clear, but only useful in the case where the filesystem isn't shared between machines (such as some CI systems, not sure about yours).

**Comment by benjamin-bader** [comment]
ccache is conceptually a hash table; it uses a hash of source-file path, mtime, compiler flags, etc. for each compiled object file.  It sits between the build system and the compiler (hence, `CMAKE_CXX_COMPILER_LAUNCHER`).  It will hash the compiler arguments, and if there is a matching object file in the cache, will just use that rather than compiling.

In constrast, `--target-path` just specifies the build output directory.  Make (and any build system, really) will avoid compiling files that haven't changed, if there is a suitably-named object file in the output directory whose mtime is greater than the source file. 

`ccache` maintains its cache elsewhere, so you can delete part (or all) of the `--target-path` and still have a fast build.  This is why CI benefits so much - builds typically happen in an isolated arena without any shared `--target-path`; with `ccache`, you can avoid recompiling things that haven't changed.

**Comment by digitalinfinity** [comment]
+1 to this PR- @obastemur what's the downside of supporting ccache along with --target-path?

**Comment by obastemur** [comment]
I don’t really see how we are going to benefit from that OR use / test that so we know when it breaks

**Comment by jackhorton** [comment]
seems like a simple enough feature that 1) it would be difficult to break and 2) even if we did break it at some point somehow, it is still preferable between now and then for embedders who want to use this to not have to fork our build system to accomplish it. Also, if we can get SCCache + Azure to work (which Benjamin has a different PR out for, I noticed), there is a world in which this sounds really great for use in Jenkins, especially if we wanted to keep adding on to our build matrix. For what its worth, is --target-path tested or used by anything right now?
